### ISSUE TITLE
Score always -1 if without early stopping

### ISSUE THREAD

**Issue by dk-forestry** [microsoft/LightGBM#2029]
## Environment info

Operating System, CPU/GPU model, C++/Python/R version, LightGBM version: Kaggle kernel

## Description

- lgb.cv + early_stopping_rounds: works as expected 
- lgb.cv + no early_stopping_rounds: seems to work, but best_score is always -1 
- lgb.train + early_stopping_rounds: throws an error (because no valids)
- lgb.train + no early_stopping_rounds: seems to work, but best_score is always -1 

Is this expected behavior? 

## Reproducible examples

Load libraries
```
library(tidyverse)
library(lightgbm)
library(Matrix)
```
Create data
```
data("bank")
train_for_pred <- as_tibble(bank)

# create target 
train_for_pred_target <- as.integer(as.factor(train_for_pred$poutcome)) - 1

# remove all character columns
train_for_pred <- train_for_pred %>%
  select_if(negate(is.character))

# create sparse matrix
train_for_pred_matrix <- as.matrix(train_for_pred)
train_for_pred_sparse <- train_for_pred_matrix %>%
  Matrix(sparse = TRUE)

train_for_pred_lgb <- lgb.Dataset(
  data = train_for_pred_sparse,
  label = train_for_pred_target
)
``` 
create parameter lists
```
para_list_with_early_stopping <- list(
  objective = "regression",
  early_stopping_rounds = 1
)

para_list_without_early_stopping <- list(
  objective = "regression"
)
```
fit models 
```
# this works as expected:
lgb.model.cv <- lgb.cv(
  params = para_list_with_early_stopping,
  data = train_for_pred_lgb,
  nfold = 5,
  verbose = 1
)
lgb.model.cv$best_score # = -0.2006221

# this seems to work, but best score is -1
lgb.model.cv <- lgb.cv(
  params = para_list_without_early_stopping,
  data = train_for_pred_lgb,
  nfold = 5,
  verbose = 1
)
lgb.model.cv$best_score # = -1

# this throws an error
model_lgb = lgb.train(params = para_list_with_early_stopping,
                      data = train_for_pred_lgb,
)
# -> Error in init(env): For early stopping, valids must have at least one element

# use lgb.unloader(wipe = TRUE) because error in previous line
lgb.unloader(wipe = TRUE)
train_for_pred_lgb <- lgb.Dataset(
  data = train_for_pred_sparse,
  label = train_for_pred_target
)

model_lgb = lgb.train(params = para_list_without_early_stopping,
                      data = train_for_pred_lgb,
)
model_lgb$best_score # = -1

```

**Comment by guolinke** [comment]
ping @Laurae2

**Comment by guolinke** [comment]
ping @Laurae2 again.
Any idea about this early stopping issue?

**Comment by Laurae2** [comment]
@guolinke I think this is set by default because without early stopping, the callback for scoring is not added (because there is nothing to score for early stopping).

https://github.com/Microsoft/LightGBM/blob/master/R-package/R/lgb.Booster.R#L8

LightGBM does not provide training metric while training, therefore the `best_iter` and `best_score` are "empty" (-1).

**Comment by StrikerRUS** [comment]
I think that usage of `train` function without validation data is expected to be resulted in undefined `best_score`. In Python wrapper the default value is `{}`:
https://github.com/Microsoft/LightGBM/blob/0e025e9c223023e49f90b1d026b6eb287d125972/python-package/lightgbm/basic.py#L1602

**Comment by Laurae2** [comment]
@StrikerRUS I just checked and the following is incorrect in the R-package:

* Evaluation is recorded with and without early stopping (with validation data)
* Best score is not recorded without early stopping

It should be fixed.

@jameslamb are you fixing it in the rework of the R-package or should we fix it before?

**Comment by jameslamb** [comment]
@Laurae2 it should be fixed before.

I also will open up a work in progress (WIP) PR in the next few days so you can keep an eye on what is changing in the rework and offer advice for mistakes I'm making (I am definitely not a C++ expert).

**Comment by StrikerRUS** [comment]
@Laurae2 Was it fixed in #2159?

**Comment by Laurae2** [comment]
@StrikerRUS Yes, this is fixed.
